std-bits/tableau/src/main/java/org/enso/tableau/HyperFormat.java

```diff
@@ -350,59 +375,204 @@ private static TableDefinition createTable(
+      Column[] sourceColumns = table.getColumns();
+      int defColumnCount = tableDef.getColumns().size();
+
+      return IntStream.range(0, defColumnCount)
+          .mapToObj(
+              i ->
+                  i < sourceColumns.length
+                      ? sourceColumns[i].getStorage()
+                      : new NullStorage(numberOfRows))
+          .toArray(ColumnStorage[]::new);
+    }
+  }
+
+  private static void addValueToInserter(Inserter inserter, ColumnStorage storage, int row) {
+    if (storage.isNothing(row)) {
+      inserter.addNull();
+    } else if (storage instanceof ColumnDoubleStorage doubleStorage) {
+      inserter.add(doubleStorage.getItemAsDouble(row));
+    } else if (storage instanceof ColumnLongStorage longStorage) {
+      inserter.add(longStorage.getItemAsLong(row));
+    } else if (storage instanceof ColumnBooleanStorage boolStorage) {
+      inserter.add(boolStorage.getItemAsBoolean(row));
+    } else {
+      Object value = storage.getItemBoxed(row);
+      switch (value) {
+        case String s -> inserter.add(s);
+        case LocalDate ld -> inserter.add(ld);
+        case LocalTime lt -> inserter.add(lt);
+        case ZonedDateTime zdt -> inserter.add(zdt);
+        case BigDecimal bd -> inserter.add(bd);
+        default -> throw new HyperUnsupportedTypeError(value.toString());
+      }
+    }
+  }
+
+  private static void validateNoExtraColumnsByName(
+      Table table,
+      String[] allowedColumnNames,
+      List<String> warningUnmatchedColumns,
+      boolean throwDontWarn) {
+    Set<String> allowed = Set.of(allowedColumnNames);
+    Set<String> tableColumnNames =
+        Arrays.stream(table.getColumns()).map(Column::getName).collect(Collectors.toSet());
+
+    String[] extraColumns =
+        tableColumnNames.stream().filter(name -> !allowed.contains(name)).toArray(String[]::new);
+
+    if (extraColumns.length > 0) {
+      throw new HyperUnmatchedColumns(extraColumns);
+    }
+
+    List<String> missingColumns =
+        Arrays.stream(allowedColumnNames).filter(name -> !tableColumnNames.contains(name)).toList();
+
+    if (!missingColumns.isEmpty()) {
+      if (throwDontWarn) {
+        throw new HyperUnmatchedColumns(missingColumns.toArray(String[]::new));
+      } else {
+        warningUnmatchedColumns.addAll(missingColumns);
+      }
+    }
+  }
+
+  private static void validateNoExtraColumnsByPosition(
+      Table table,
+      TableDefinition tableDef,
+      List<String> warningUnmatchedColumns,
+      boolean throwDontWarn) {
+    int tableColumnCount = table.getColumns().length;
+    int defColumnCount = tableDef.getColumns().size();
+
+    if (tableColumnCount > defColumnCount) {
+      String[] extraColumnNames =
+          IntStream.range(defColumnCount, tableColumnCount)
+              .mapToObj(i -> table.getColumns()[i].getName())
+              .toArray(String[]::new);
+
+      throw new HyperUnmatchedColumns(extraColumnNames);
```

Comment:
This feels like the wrong error - we should report a mismatched column count.
We have a `Column_Count_Mismatch` error in Standard.Table.

Reply:
This get's tricky in the warning case though. I'm tempted to leave this one.